statvar_imports/school_retention/download_script.py

```diff
@@ -0,0 +1,267 @@
+    
+    if "LEAID" in df.columns:
+        # Remove non-digits, then pad to 7 (e.g., "100005" -> "0100005")
+        df["LEAID"] = clean_id_column(df["LEAID"]).str.zfill(7)
+    
+    if "SCHID" in df.columns:
+        # Remove non-digits, then pad to 5 (e.g., "879" -> "00879")
+        df["SCHID"] = clean_id_column(df["SCHID"]).str.zfill(5)
+        
+    if "COMBOKEY" in df.columns:
+        df["COMBOKEY"] = clean_id_column(df["COMBOKEY"])
+
+    # --- 2. GENERATE NCESID (STRICT LOGIC) ---
+    # Logic: Ignore the source COMBOKEY because it is often corrupt (missing zeros).
+    # Instead, construct it: NCESID = LEAID (7) + SCHID (5)
+    
+    if "LEAID" in df.columns and "SCHID" in df.columns:
+        df["NCESID"] = df["LEAID"] + df["SCHID"]
+    elif "COMBOKEY" in df.columns:
+        # Fallback ONLY if LEAID or SCHID is missing
+        logging.warning(f"Using raw COMBOKEY fallback for {filename}")
+        df["NCESID"] = df["COMBOKEY"].str.zfill(12)
+    else:
+        logging.warning(f"Missing LEAID/SCHID or COMBOKEY in {filename}")
+        df["NCESID"] = ""
+
+    # --- 3. CLEAN YES/NO ---
+    exclude_upper = [c.upper() for c in EXCLUDE_COLS]
+    # target columns are those NOT in the exclude list and NOT observationDate
+    target_cols = [col for col in df.columns if col not in exclude_upper and col != 'observationDate']
+    
+    # Replace Yes/No with 1/0
+    df[target_cols] = df[target_cols].replace(REPLACEMENT_MAP)
+    
+    # Convert numeric columns, turning errors into NaN (blank)
+    df[target_cols] = df[target_cols].apply(pd.to_numeric, errors='coerce')
+
+    # --- 4. REORDER COLUMNS ---
+    priority_cols = ['observationDate', 'NCESID']
+    existing_priority = [c for c in priority_cols if c in df.columns]
+    other_cols = [c for c in df.columns if c not in existing_priority]
+    
+    return df[existing_priority + other_cols]
+
+def process_dataframe(file_path: str, year: int):
+    """
+    Loads, cleans, and saves ONLY the data sheet.
+    """
+    try:
+        filename = os.path.basename(file_path)
+
+        # --- CSV HANDLING ---
+        if file_path.lower().endswith(".csv"):
+            try:
+                df = pd.read_csv(file_path, dtype=str, low_memory=False, encoding='latin1')
+            except UnicodeDecodeError:
+                df = pd.read_csv(file_path, dtype=str, low_memory=False)
+
+            df = transform_crdc_data(df, year, filename)
+            df.to_csv(file_path, index=False)
+
+        # --- XLSX HANDLING ---
+        elif file_path.lower().endswith(".xlsx"):
+            all_sheets = pd.read_excel(file_path, dtype=str, sheet_name=None)
+            
+            # Identify the Data Sheet (usually the first one)
+            data_sheet_name = list(all_sheets.keys())[0]
+            df = all_sheets[data_sheet_name]
+
+            df = transform_crdc_data(df, year, filename)
+
+            # Save only the DATA sheet
+            with pd.ExcelWriter(file_path) as writer:
```

Comment:
Pls don't modify the original source data as it was downloaded. Instead generate another file with a prefix such as transformed_input*